controller/internal/enforcer/nfqdatapath/nfq_linux.go

```diff
@@ -75,20 +80,44 @@ func (d *Datapath) processNetworkPacketsFromNFQ(p *nfqueue.NFPacket) {
 
 	// Parse the packet - drop if parsing fails
 	netPacket, err := packet.New(packet.PacketTypeNetwork, p.Buffer, strconv.Itoa(p.Mark), true)
-
+	var processError error
+	var tcpConn *connection.TCPConnection
+	var udpConn *connection.UDPConnection
 	if err != nil {
 		netPacket.Print(packet.PacketFailureCreate)
 	} else if netPacket.IPProto == packet.IPProtocolTCP {
-		err = d.processNetworkTCPPackets(netPacket)
+		tcpConn, processError = d.processNetworkTCPPackets(netPacket)
 	} else if netPacket.IPProto == packet.IPProtocolUDP {
-		err = d.ProcessNetworkUDPPacket(netPacket)
+		udpConn, processError = d.ProcessNetworkUDPPacket(netPacket)
 	} else {
-		err = fmt.Errorf("invalid ip protocol: %d", netPacket.IPProto)
+		processError = fmt.Errorf("invalid ip protocol: %d", netPacket.IPProto)
+
 	}
-	if err != nil {
+
+	if processError != nil {
 		length := uint32(len(p.Buffer))
 		buffer := p.Buffer
 		p.QueueHandle.SetVerdict2(uint32(p.QueueHandle.QueueNum), 0, uint32(p.Mark), length, uint32(p.ID), buffer)
+		if netPacket.IPProto == packet.IPProtocolTCP {
+			d.collectTCPPacket(&debugpacketmessage{
+				Mark:    p.Mark,
+				p:       netPacket,
+				tcpConn: tcpConn,
+				udpConn: nil,
+				err:     processError,
+				network: true,
+			})
+		} else if netPacket.IPProto == packet.IPProtocolUDP {
+			d.collectUDPPacket(&debugpacketmessage{
+				Mark:    p.Mark,
+				p:       netPacket,
+				tcpConn: nil,
+				udpConn: udpConn,
+				err:     processError,
+				network: true,
+			})
+		}
+
 		return
```

Comment:
I think we can make a function for this and call it from a couple of places. Also, we only report these packets if we have tracing on. The performance penalty is too high otherwise.